File: 53-maximum-subarray/53-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
// var maxSubArray = function(nums) {
//     let maxSum = -Infinity
//     let i = 0
    
//     while (i < nums.length) {
       
//     }
//     return maxSum
// };


var maxSubArray = function(nums) {
    let max = -Infinity

    let currSum = 0
    for (let num of nums) {
        currSum += num
        if (currSum < 0) {
            max = Math.max(currSum, max)
            currSum = 0
        } else {
            max = Math.max(currSum, max)
        }
    }
    return max
}

File: 53-maximum-subarray/53-maximum-subarray.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        max_sum = float('-inf')
        
        sum = 0
        
        for num in nums:
            sum += num
            if max_sum <= sum:
                max_sum = sum
            if sum < 0:
                sum = 0
        
        return max_sum
                
        